src/uu/tail/src/tail.rs

```diff
@@ -23,537 +23,68 @@ extern crate clap;
-
-        // Mimic GNU's tail for -[nc]0 without -f and exit immediately
-        if settings.follow.is_none() && !starts_with_plus && {
-            if let FilterMode::Lines(l, _) = settings.mode {
-                l == 0
-            } else {
-                settings.mode == FilterMode::Bytes(0)
-            }
-        } {
-            std::process::exit(0)
-        }
 
-        settings.retry =
-            matches.contains_id(options::RETRY) || matches.contains_id(options::FOLLOW_RETRY);
-
-        if settings.retry && settings.follow.is_none() {
-            show_warning!("--retry ignored; --retry is useful only when following");
-        }
-
-        if matches.contains_id(options::ZERO_TERM) {
-            if let FilterMode::Lines(count, _) = settings.mode {
-                settings.mode = FilterMode::Lines(count, 0);
-            }
-        }
-
-        settings.stdin_is_pipe_or_fifo = matches.contains_id(options::PRESUME_INPUT_PIPE);
-
-        settings.paths = matches
-            .get_many::<String>(options::ARG_FILES)
-            .map(|v| v.map(PathBuf::from).collect())
-            .unwrap_or_default();
-
-        settings.verbose = (matches.contains_id(options::verbosity::VERBOSE)
-            || settings.paths.len() > 1)
-            && !matches.contains_id(options::verbosity::QUIET);
-
-        Ok(settings)
-    }
-
-    fn follow_descriptor(&self) -> bool {
-        self.follow == Some(FollowMode::Descriptor)
-    }
-
-    fn follow_name(&self) -> bool {
-        self.follow == Some(FollowMode::Name)
-    }
-
-    fn follow_descriptor_retry(&self) -> bool {
-        self.follow_descriptor() && self.retry
-    }
+use uucore::display::Quotable;
+use uucore::error::{get_exit_code, set_exit_code, FromIo, UError, UResult, USimpleError};
 
-    fn follow_name_retry(&self) -> bool {
-        self.follow_name() && self.retry
-    }
-}
+pub use args::uu_app;
+use args::{parse_args, FilterMode, Settings, Signum};
+use chunks::ReverseChunks;
+use follow::WatcherService;
+use paths::{FileExtTail, Input, InputKind, InputService, MetadataExtTail};
 
 #[uucore::main]
 pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-    let matches = uu_app().try_get_matches_from(arg_iterate(args)?)?;
-    let mut settings = Settings::from(&matches)?;
-
-    // skip expensive call to fstat if PRESUME_INPUT_PIPE is selected
-    if !settings.stdin_is_pipe_or_fifo {
-        settings.stdin_is_pipe_or_fifo = stdin_is_pipe_or_fifo();
-    }
-
-    uu_tail(settings)
+    let settings = parse_args(args)?;
+    uu_tail(&settings)
 }
 
-fn uu_tail(mut settings: Settings) -> UResult<()> {
-    let dash = PathBuf::from(text::DASH);
-
+fn uu_tail(settings: &Settings) -> UResult<()> {
     // Mimic GNU's tail for `tail -F` and exit immediately
-    if (settings.paths.is_empty() || settings.paths.contains(&dash)) && settings.follow_name() {
+    let mut input_service = InputService::from(settings);
+    let mut watcher_service = WatcherService::from(settings);
```

Comment:
I find the `InputService` and `WatcherService` to be a little bit vague. `Service` could mean basically anything.

Reply:
No problem. That's acutally one of the reasons, why I like using Service as a placeholder for such structures in the beginning of writing. I'll figure out something more meaningful and descriptive when I'm finishing this up and the purpose of these structures is clearified.